### PR TITLE
Feat generalized ability

### PR DESCRIPTION
**Pull request by mboedigh**
This feature adds a switch to allow client to get available specific ability ids for a given unit or units. 
This allows the creation of an ability graph containing all possible abilities in the game (thus everything a bot could do) without forming cycles. A bot can discover the entire game.

Sorry for multiple commits. I was unable squash them.

### CODE REVIEW

**Comment by alkurbatov** [comment]
Sorry, but what does 'specific ability ids' means? Could you please provide an example?

File: include/sc2api/sc2_interfaces.h [hunk 1]
```diff
@@ -284,9 +283,8 @@ class QueryInterface {
 
     struct PlacementQuery {
         PlacementQuery() = default;
-        PlacementQuery(AbilityID ability_id, Point2D target) :
-            ability(ability_id),
-            target_pos(target) {};
+        PlacementQuery(AbilityID ability_id, Point2D target) : ability(ability_id),
+                                                               target_pos(target){};
```
Comment (alkurbatov) [inline include/sc2api/sc2_interfaces.h]: Please, leave it as is.

Comment (mboedigh) [reply include/sc2api/sc2_interfaces.h]: Sorry, I missed that. It should be fixed. 
It was VS formatting what it thinks is Google.

**Comment by mboedigh** [comment]
Here is an example of what I'm talking about.
`sc2::AbilityData` defines ability_id and remaps_to_ability_id. The first is specific and the second is general.  
You see these in typeenums too. For example
```
RESEARCH_TERRANINFANTRYARMOR = 3697,  // General
RESEARCH_TERRANINFANTRYARMORLEVEL1 = 656,   // Specific
```

By default the game uses general_id for ability ids seen in `unit->orders`.  This can be configured in the coordinator to use specific ids. But unfortunately, it isn't yet possible to get a list of specific  abilities available for a unit.

The value in using specific ids is that a graph can be constructed (e.g. by a bot) that can tell the there are 3 upgrade levels and they are all different. Using general ids would create a cycle
> Armor Upgrade -> Armor Upgrade -> … -> Armor Upgrade. 

While using specific Ids gives a unique graph. 
> armor I -> armor II -> armor III. 

If you would rather discuss new ideas in the discussion prior to pull-requests let me know.

**Comment by alkurbatov** [comment]
Thank you for good explanation. It is clear now.

Regarding the PR: by some reason contrib/protobuf was changed too. I think it shouldn't.

File: include/sc2api/sc2_interfaces.h [hunk 2]
```diff
@@ -245,13 +244,13 @@ class QueryInterface {
...
     //!< \return Abilities for the unit.
-    virtual AvailableAbilities GetAbilitiesForUnit(const Unit* unit, bool ignore_resource_requirements = false) = 0;
+    virtual AvailableAbilities GetAbilitiesForUnit(const Unit* unit, bool ignore_resource_requirements = false, bool use_generalized_ability = true) = 0;
     //! Issues multiple available abilities queries.
     //! Batch version.
     //!< \param tag Tags of units.
     //!< \param ignore_resource_requirements Ignores food, mineral and gas costs, as well as cooldowns.
     //!< \return Abilities for the units.
-    virtual std::vector<AvailableAbilities> GetAbilitiesForUnits(const Units& units, bool ignore_resource_requirements = false) = 0;
+    virtual std::vector<AvailableAbilities> GetAbilitiesForUnits(const Units& units, bool ignore_resource_requirements = false, bool use_generalized_ability = true) = 0;
```
Comment (alkurbatov) [inline include/sc2api/sc2_interfaces.h]: Please, document the new parameter.

Comment (alkurbatov) [reply include/sc2api/sc2_interfaces.h]: Here and above.

**Comment by mboedigh** [comment]
I don’t think I changed protobuf. I think I saw it in a commit I pulled merging your dev branch. I haven’t noticed any problems with it however.  Is there something I should do to purge it?

**Comment by alkurbatov** [comment]
Yes, it was in one my recent commits.
You short exclude this change from the original commit as it is not related to the provided feature.

**Comment by mboedigh** [comment]
I made changes to document the new parameters. 
As for the submodule change. ¯\_(ツ)_/¯ Again, over my head. :). 

Here is my local git log

```
ab65e63 (HEAD -> FeatGeneralizedAbility) feat: Allow retrieving specific ability ids with GetAbilitiesFromUnit(s)
65907e1 (upstream/dev) Merge pull request #56 from Cryptyc/dev
f27c954 feat: Added legacy typeenums for compatibility
bed887a build: Improve compilation under VS 2019
2ae5519 feat: Upgraded type enums for 4.11.3
cc3b534 feat: add use_generalized_ability option when retrieving abilities using GetAbilitiesForUnit/s
52e87b0 chore: Extend .editorconfig
f90fcc6 Merge pull request #53 from mboedigh/FixMissingTypeEnums
```

The protobuf change was in bed8. Are you asking that I rebase f27c on 2ae5 and skip bed8 and then force push?

**Comment by alkurbatov** [comment]
The problem here is that you've reverted my commit.
This is in my commit:
`Submodule protobuf updated from f56542 to 1a74ba`

This is in your PR:
`Submodule protobuf updated from 1a74ba to f56542`

Of course it shouldn't happen.

**Pull request merged**